device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java

```diff
@@ -229,7 +255,26 @@ public DeviceClient(String connString, IotHubClientProtocol protocol, String pub
      */
     public DeviceClient(String connString, IotHubClientProtocol protocol, SSLContext sslContext) throws URISyntaxException
     {
-        super(new IotHubConnectionString(connString), protocol, sslContext, SEND_PERIOD_MILLIS, getReceivePeriod(protocol));
+        super(new IotHubConnectionString(connString), protocol, sslContext, SEND_PERIOD_MILLIS, getReceivePeriod(protocol), null);
+        commonConstructorVerifications();
+        commonConstructorSetup();
+    }
+
+    /**
+     * Creates a device client that uses the provided SSLContext for SSL negotiation
+     * @param connString the connection string for the device. May be an x509 connection string (format: "HostName=...;DeviceId=...;x509=true")
+     *                   and it may be a SAS connection string (format: "HostName=...;DeviceId=...;SharedAccessKey=..."). If
+     *                   this connection string is an x509 connection string, the client will use the provided SSLContext for authentication.
+     * @param protocol the protocol to use when communicating with IotHub
+     * @param sslContext the ssl context that will be used during authentication. If the provided connection string does not contain
+     *                   SAS based credentials, then the sslContext will be used for x509 authentication. If the provided connection string
+     *                   does contain SAS based credentials, the sslContext will still be used during SSL negotiation.
+     * @param clientOptions The options that allow configuration of the device client instance during initialization
+     * @throws URISyntaxException if the hostname in the connection string is not a valid URI
+     */
+    public DeviceClient(String connString, IotHubClientProtocol protocol, SSLContext sslContext, ClientOptions clientOptions) throws URISyntaxException
```

Comment:
Why can't SSLContext be part of clientOptions?

Reply:
How would that work with other constructors?

Reply:
Is it possible to just have 2 constructors and maybe SSLContext can be set to null if user does not want to use it?

Reply:
As this is a new API, I don't see reason why you can't do that.

Reply:
We have resolved this offline and I will be sending out a PR for these changes.